[Quant] Add qint32 quantization support

terrychenism opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72472

Summary:
Add dtype=int32 support for observer

Test Plan:
python3 test/test_quantization.py TestObserver.test_per_tensor_observers

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34056640](https://our.internmc.facebook.com/intern/diff/D34056640)

terrychenism committed:
[Quant] Add qint32 quantization support

Summary:
Add dtype=int32 support for observer

Test Plan:
python3 test/test_quantization.py TestObserver.test_per_tensor_observers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34056640).

vkuzo commented on torch/ao/quantization/utils.py:
```diff
@@ -234,7 +234,10 @@ def calculate_qmin_qmax(quant_min: int, quant_max: int, has_customized_qrange: b
         # The actual values of initial_qmin and initial_qmax will be reset below.
-        initial_quant_min, initial_quant_max = 0, 255
+        if dtype == torch.qint32:
+            initial_quant_min, initial_quant_max = 0, 2**32 - 1
```
should these values be for signed int?

terrychenism replied:
> should we set default value as 2**31 for signed int?

jerryzh168 commented on test/quantization/core/test_workflow_module.py:
```diff
@@ -106,12 +107,20 @@ def test_per_tensor_observers(self, qdtype, qscheme, reduce_range):
-                    ref_zero_point = 0 if qdtype is torch.qint8 else 128
+                if qdtype == torch.qint32:
+                    if qscheme == torch.per_tensor_symmetric:
+                        ref_scale = 0.000244
```
is this calculated from some formula?

terrychenism replied:
yes, formula from this function: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/observer.py#L258

terrychenism replied:
formula from this function https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/observer.py#L258

jerryzh168 replied:
I mean should we write down the formula instead of a constant value?

jerryzh168 commented on torch/ao/quantization/utils.py:
```diff
@@ -245,9 +248,14 @@ def calculate_qmin_qmax(quant_min: int, quant_max: int, has_customized_qrange: b
+                0 < qrange_len <= 2**32
+            ), "quantization range should be positive and not exceed the maximum bit range (=4294967296)."
         if dtype == torch.qint8:
             quant_min, quant_max = -qrange_len // 2, qrange_len // 2 - 1
```
why do we need to get the qmin/qmax based on qrange_len?

terrychenism commented on torch/ao/quantization/utils.py:
```diff
@@ -245,9 +248,14 @@ def calculate_qmin_qmax(quant_min: int, quant_max: int, has_customized_qrange: b
         qrange_len = initial_quant_max - initial_quant_min + 1
-        assert (
-            0 < qrange_len <= 256
-        ), "quantization range should be positive and not exceed the maximum bit range (=256)."
+        if dtype == torch.qint8:
+            assert (
+                0 < qrange_len <= 256
+            ), "quantization range should be positive and not exceed the maximum bit range (=256)."
+        elif dtype == torch.qint32:
+            assert (
+                0 < qrange_len <= 2**32
+            ), "quantization range should be positive and not exceed the maximum bit range (=4294967296)."
```
ok, then we don't need qrange_len?

terrychenism replied:
Yes, then we don't need qrange_len, right?

jerryzh168 replied:
yeah I'm not sure why we need it.. we can try removing it I think, but it could be in a separate PR

terrychenism committed:
Update on "[Quant] Add qint32 quantization support"


Summary:
Add dtype=int32 support for observer

Test Plan:
python3 test/test_quantization.py TestObserver.test_per_tensor_observers

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34056640](https://our.internmc.facebook.com/intern/diff/D34056640)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34056640).

terrychenism committed:
Update on "[Quant] Add qint32 quantization support"


Summary:
Add dtype=int32 support for observer

Test Plan:
python3 test/test_quantization.py TestObserver.test_per_tensor_observers

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34056640](https://our.internmc.facebook.com/intern/diff/D34056640)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34056640).

terrychenism committed:
Update on "[Quant] Add qint32 quantization support"


Summary:
Add dtype=int32 support for observer

Test Plan:
python3 test/test_quantization.py TestObserver.test_per_tensor_observers

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34056640](https://our.internmc.facebook.com/intern/diff/D34056640)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34056640).

terrychenism committed:
Update on "[Quant] Add qint32 quantization support"


Summary:
Add dtype=int32 support for observer

Test Plan:
python3 test/test_quantization.py TestObserver.test_per_tensor_observers

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34056640](https://our.internmc.facebook.com/intern/diff/D34056640)

[ghstack-poisoned]

terrychenism committed:
Update on "[Quant] Add qint32 quantization support"


Summary:
Add dtype=int32 support for observer

Test Plan:
python3 test/test_quantization.py TestObserver.test_per_tensor_observers

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34056640](https://our.internmc.facebook.com/intern/diff/D34056640)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34056640).

terrychenism committed:
Update on "[Quant] Add qint32 quantization support"


Summary:
Add dtype=int32 support for observer

Test Plan:
python3 test/test_quantization.py TestObserver.test_per_tensor_observers

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34056640](https://our.internmc.facebook.com/intern/diff/D34056640)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34056640).
